subsys/bluetooth/audio/tbs.c

```diff
@@ -1499,58 +1508,211 @@ static void signal_interval_timeout(struct k_work *work)
 {
-	gtbs_service_inst_init(&gtbs_inst, &gtbs_svc);
+	size_t str_len;
 
-	for (size_t i = 0; i < ARRAY_SIZE(svc_insts); i++) {
-		int err;
+	if (param == NULL) {
+		LOG_DBG("param is NULL");
 
-		err = bt_gatt_service_register(&tbs_service_list[i]);
-		if (err != 0) {
-			LOG_ERR("Could not register TBS[%d]: %d", i, err);
+		return false;
+	}
+
+	if (param->provider_name == NULL) {
+		LOG_DBG("provider_name is NULL");
+
+		return false;
+	}
+
+	str_len = strlen(param->provider_name);
+	if (str_len > CONFIG_BT_TBS_MAX_PROVIDER_NAME_LENGTH) {
+		LOG_DBG("Provider name length (%zu) larger than "
+			"CONFIG_BT_TBS_MAX_PROVIDER_NAME_LENGTH %d",
+			str_len, CONFIG_BT_TBS_MAX_PROVIDER_NAME_LENGTH);
+
+		return false;
+	}
+
+	if (param->uci == NULL) {
+		LOG_DBG("uci is NULL");
+
+		return false;
+	}
+
+	if (param->uri_schemes_supported == NULL) {
+		LOG_DBG("uri_schemes_supported is NULL");
+
+		return false;
+	}
+
+	if (!IN_RANGE(param->technology, BT_TBS_TECHNOLOGY_3G, BT_TBS_TECHNOLOGY_WCDMA)) {
+		LOG_DBG("Invalid technology: %u", param->technology);
+
+		return false;
+	}
+
+	if (param->supported_features > BT_TBS_FEATURE_ALL) {
+		LOG_DBG("Invalid supported_features: %u", param->supported_features);
+
+		return false;
+	}
+
+	if (CONFIG_BT_TBS_BEARER_COUNT == 0 && !param->gtbs) {
+		LOG_DBG("Cannot register TBS when CONFIG_BT_TBS_BEARER_COUNT=0");
+
+		return false;
+	}
+
+	return true;
+}
+
+int bt_tbs_register_bearer(const struct bt_tbs_register_param *param)
+{
+	int ret;
+
+	CHECKIF(!valid_register_param(param)) {
+		LOG_DBG("Invalid parameters");
+
+		return -EINVAL;
+	}
+
+	if (param->gtbs && inst_is_registered(&gtbs_inst)) {
+		LOG_DBG("GTBS already registered");
+
+		return -EALREADY;
+	}
+
+	if (!param->gtbs && !inst_is_registered(&gtbs_inst)) {
+		LOG_DBG("GTBS not yet registered");
+
+		return -EAGAIN;
+	}
+
+	if (param->gtbs) {
+		ret = gtbs_service_inst_register(param);
+		if (ret < 0) {
+			LOG_DBG("Failed to register GTBS: %d", ret);
+
+			return -ENOEXEC;
 		}
+	} else if (CONFIG_BT_TBS_BEARER_COUNT > 0) {
+		ret = tbs_service_inst_register(param);
+		if (ret < 0) {
+			LOG_DBG("Failed to register GTBS: %d", ret);
 
-		tbs_service_inst_init(&svc_insts[i], &tbs_service_list[i]);
+			if (ret == -ENOMEM) {
+				return -ENOMEM;
+			}
+
+			return -ENOEXEC;
+		}
 	}
 
-	return 0;
+	/* ret will contain the index of the registered service */
+	return ret;
 }
 
-SYS_INIT(bt_tbs_init, APPLICATION, CONFIG_KERNEL_INIT_PRIORITY_DEVICE);
+int bt_tbs_unregister_bearer(uint8_t bearer_index)
+{
+	struct tbs_inst *inst = inst_lookup_index(bearer_index);
+	struct bt_gatt_service *svc;
+	struct k_work_sync sync;
+	bool restart_reporting_interval;
+	int err;
+
+	if (inst == NULL) {
+		LOG_DBG("Could not find inst by index %u", bearer_index);
+
+		return -EINVAL;
+	}
+
+	if (!inst_is_registered(inst)) {
+		LOG_DBG("Instance from index %u is not registered", bearer_index);
+
+		return -EALREADY;
+	}
+
+	if (inst_is_gtbs(inst)) {
```

Comment:
If this is the GTBS, we need to check that all ordinary instances are unregistered before we unregister this, right? Doesn't seem like we are doing that right now.